File: Java/Datastructure/Stack/StackUsingLinkedList.java
package com.company;

public class StackUsingLinkedList {
    Node root;

    public static class Node{
        int data;
        Node next;
        Node(int d){
           data = d;
            next=null;
        }
    }

    public boolean isEmpty(){
        return root==null;
    }

    public void push(int data){
        Node newNode = new Node(data);
        if (root==null){
            root=newNode;
        }else {
            Node temp = root;
            root = newNode;
            newNode.next=temp;
        }
    }

    public int pop(){
        int popped = Integer.MIN_VALUE;
        if (root==null){
            System.out.println("Stack is empty");
        }else {

            popped=root.data;
            root=root.next;
        }
        return popped;
    }

    public int peek(){
        if (root==null){
            System.out.println("Stack is empty");
            return Integer.MIN_VALUE;
        }else {
            return root.data;
        }
    }


    public static void main(String[] args) {
        StackUsingLinkedList sll = new StackUsingLinkedList();
//        System.out.println(sll.pop());
        sll.push(10);
        sll.push(20);
        sll.push(30);

        System.out.println(sll.pop() + " is popped");

        System.out.println("top element is "+ sll.peek());


    }
}

File: stack_using_linked_list.cpp
#include<bits/stdc++.h>
using namespace std;
template<typename T>    // for choice of data type
class Node{
    public:
    T data;
    Node* next;

    Node(T data){
        this->data=data;
        next=NULL;
    }
};

template<typename T>
class StackUsingLL{
    Node<T>* head;
    int count;

    public:
    StackUsingLL(){
        head=NULL;
        count=0;
    }
    int getSize(){
        return count;
    }
    bool isEmpty(){
        return count==0;
    }
    void push(T element){
        Node<T>*newNode=new Node<T>(element);
        newNode->next=head;
        head=newNode;
        count++;
    }
    T pop(){
        if(isEmpty()){
            cout<<"stack empty"<<endl;
            return 0;
        }
        T returnValue=head->data;
        Node<T>*temp=head;
        head=head->next;
        delete temp;
        count--;
        return returnValue;
    }
    T top(){
        if(isEmpty()){
            cout<<"stack empty"<<endl;
            return 0;
        }
        return head->data;
    }
};

int main(){
    StackUsingLL<int>s;
    int arr[]={1,2,3,4};
    for(int i=0;i<4;i++){
        s.push(arr[i]);
    }
    for(int i=0;i<4;i++){
        cout<<s.pop()<<endl;
    }
}